src/Symfony/Component/Mailer/Bridge/Sendgrid/Webhook/SendgridRequestParser.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Mailer\Bridge\Sendgrid\Webhook;
+
+use EllipticCurve\Ecdsa;
+use EllipticCurve\PublicKey;
+use EllipticCurve\Signature;
+use Symfony\Component\DependencyInjection\Exception\LogicException;
+use Symfony\Component\HttpFoundation\ChainRequestMatcher;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestMatcher\IsJsonRequestMatcher;
+use Symfony\Component\HttpFoundation\RequestMatcher\MethodRequestMatcher;
+use Symfony\Component\HttpFoundation\RequestMatcherInterface;
+use Symfony\Component\Mailer\Bridge\Sendgrid\RemoteEvent\SendgridPayloadConverter;
+use Symfony\Component\RemoteEvent\Event\Mailer\AbstractMailerEvent;
+use Symfony\Component\RemoteEvent\Exception\ParseException;
+use Symfony\Component\Webhook\Client\AbstractRequestParser;
+use Symfony\Component\Webhook\Exception\RejectWebhookException;
+
+/**
+ * @author WoutervanderLoop.nl <info@woutervanderloop.nl>
+ */
+final class SendgridRequestParser extends AbstractRequestParser
+{
+    public function __construct(
+        private readonly SendgridPayloadConverter $converter,
+    ) {
+    }
+
+    protected function getRequestMatcher(): RequestMatcherInterface
+    {
+        return new ChainRequestMatcher([
+            new MethodRequestMatcher('POST'),
+            new IsJsonRequestMatcher(),
+        ]);
+    }
+
+    protected function doParse(Request $request, string $secret): ?AbstractMailerEvent
+    {
+        $content = $request->toArray();
+        if (
+            !isset($content[0]['email'])
+            || !isset($content[0]['timestamp'])
+            || !isset($content[0]['event'])
+            || !isset($content[0]['sg_message_id'])
+        ) {
+            throw new RejectWebhookException(406, 'Payload is malformed.');
+        }
+
+        if ($request->headers->get('X-Twilio-Email-Event-Webhook-Signature')
+            && $request->headers->get('X-Twilio-Email-Event-Webhook-Timestamp')
+        ) {
```

Comment:
Shouldn't we add a configurable option to enforce signature expectation? The current implementation exposes a security vulnerability because it allows the client to make a malformed request without the headers, causing the message to be processed even though a signature was expected, provided the signed event functionality was also enabled in SendGrid.

Reply:
Agreed, will work on that

Reply:
@SherinBloemendaal  Done